### ISSUE TITLE
torch.sum mismatch on cuda after extending a non-reduced dimension

### ISSUE THREAD

**Issue by ehsansaleh** [pytorch/pytorch#66748]
## 🐛 Bug

First of all, thank you so much for taking the time to review this report. We truly appreciate all your work. 

* Let's say I have two arrays `X` and `Y` with the shapes `X.shape = (10, u, v)` and `Y.shape = (20, u, v)`
* The first 10 rows of `X` and `Y` are **exactly** the same. In other words, you can pass the `assert torch.all(X == Y[:10])` assertion.
* Now, you evaluate `torch.sum` on the last two dimensions.
    * `X_sum = torch.sum(X, dim=(1,2))`
    * `Y_sum = torch.sum(Y, dim=(1,2))`
*  Since `X` and `Y[:10]` are exactly the same, you would expect that `X_sum` and `Y_sum[:10]` should also be exactly the same. *However, my example shows that there can be a slight mismatch*. 
## To Reproduce

1. Please download the input data numpy file `sum_bug_data.npz` from [Box](https://uofi.box.com/s/7w3ycfy2d96abifo60xuwivdcjea11ad) and place it somewhere on your disk.
    * I'm sorry that the file is huge (i.e., 60 MB), but this is the smallest file I could generate to produce such an issue.
2. Open a python terminal in the same folder.
3. Run the following script:
```
import numpy as np
import torch
npz_file = np.load('sum_bug_data.npz')
A, B = npz_file['A'], npz_file['B']

X_np = A
Y_np = np.concatenate([A, B], axis=0)
X = torch.from_numpy(X_np).cuda()
Y = torch.from_numpy(Y_np).cuda()

assert A.shape == (10, 7923, 100)
assert B.shape == (10, 7923, 100)
assert X.shape == (10, 7923, 100)
assert Y.shape == (20, 7923, 100)

X_sum = torch.sum(X, dim=(1,2)) 
Y_sum = torch.sum(Y, dim=(1,2)) 
print('X     ==     Y[:10] -> ', torch.all(X     ==     Y[:10]))
print('X_sum == Y_sum[:10] -> ', torch.all(X_sum == Y_sum[:10]))
print('-'*80)
with np.printoptions(precision=10):
    print('X_sum[:10] ->\n', X_sum[:10].detach().cpu().numpy().reshape(-1))
    print('Y_sum[:10] ->\n', Y_sum[:10].detach().cpu().numpy().reshape(-1))
    print('-'*80)
    print('X_sum[:10] == Y_sum[:10] ->\n', (X_sum[:10] == Y_sum[:10]))
```

## Expected behavior

Since `X` and `Y[:10]` are exactly the same, you would expect that `X_sum` and `Y_sum[:10]` should also be exactly the same. *However, my example shows that there can be a slight mismatch*.  In particular, there is a slight mismatch in the elements `1.7900770e+32` and `1.7900768e+32`.  

```
X     ==     Y[:10] ->  tensor(True, device='cuda:0')
X_sum == Y_sum[:10] ->  tensor(False, device='cuda:0')
--------------------------------------------------------------------------------
X_sum[:10] ->
 [8.4447355e+09 7.5724438e+26 1.1327815e+19 1.7900768e+32 9.9139092e+10
  2.0154725e+16           inf 1.8050868e+18 3.4944974e+22 6.0265007e+33]
Y_sum[:10] ->
 [8.4447355e+09 7.5724438e+26 1.1327815e+19 1.7900770e+32 9.9139092e+10
  2.0154722e+16           inf 1.8050865e+18 3.4944976e+22 6.0265007e+33]
--------------------------------------------------------------------------------
X_sum[:10] == Y_sum[:10] ->
 tensor([ True,  True,  True, False,  True, False,  True, False, False,  True],
       device='cuda:0')
```
* As you can see, there are too many `False` entries in `X_sum[:10] == Y_sum[:10]`
  * Don't mind the `inf` elements. Let's focus on the other false equivalences.

## Environment

```
PyTorch version: 1.8.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 4.8.2
Clang version: 3.4.2 (tags/RELEASE_34/dot2-final)
CMake version: version 2.8.12.2
Libc version: glibc-2.17

Python version: 3.8.8 (default, Feb 24 2021, 21:46:12)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.6.1.el7.x86_64-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Quadro RTX 6000
GPU 1: Quadro RTX 6000

Nvidia driver version: 470.74
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.8.1
[pip3] torchaudio==0.8.0a0+e4e171a
[pip3] torchvision==0.9.1
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              hfd86e86_1
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2020.2                      256
[conda] mkl-service               2.3.0            py38he904b0f_0
[conda] mkl_fft                   1.3.0            py38h54f3939_0
[conda] mkl_random                1.1.1            py38h0573a6f_0
[conda] numpy                     1.19.2           py38h54aff64_0
[conda] numpy-base                1.19.2           py38hfa32c7d_0
[conda] pytorch                   1.8.1           py3.8_cuda10.2_cudnn7.6.5_0    pytorch
[conda] torchaudio                0.8.1                      py38    pytorch
[conda] torchvision               0.9.1                py38_cu102    pytorch
```

**Comment by ngimel** [comment]
This is expected behavior, pytorch doesn't guarantee bitwise accurate results when operating on slices of tensors compared to full tensors https://pytorch.org/docs/master/notes/numerical_accuracy.html
